### PR TITLE
fix(kds): fix retry on NACK and add backoff

### PR DESCRIPTION
**Pull request by slonka**
probably the solution for root cause of https://github.com/kumahq/kuma/pull/9730#issue-2208386867

@jakubdyszkiewicz:

The current implementation only retried NACK once and it was not guaranteed.
This implementation fixes NACK retries and introduces backoff of 5s.

See comments on retry forcer for details.

This still requires changes for event-based watchdog to trigger Reconcile because in oppose to the regular watchdog, there can be a long delay until reconciller is triggered.

### Checklist prior to review

- [X] [Link to relevant issue][1] as well as docs and UI issues --
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Review by Automaat — APPROVED** [review]

**Comment by michaelbeaumont** [comment]
How does this relate to what happened in https://github.com/kumahq/kuma/issues/2164?

**Comment by slonka** [comment]
@michaelbeaumont it was introduced here https://github.com/kumahq/kuma/pull/5894 and then reverted https://github.com/kumahq/kuma/pull/5916. For now I just wanted to see if this fixes our e2e problems.

File: pkg/kds/v2/reconcile/reconciler.go [hunk 1]
```diff
@@ -116,12 +116,12 @@ func (r *reconciler) Version(new, old envoy_cache.ResourceSnapshot) (envoy_cache
...
-			changed = true
 		}
 		if new.GetVersion(typ) == version {
 			continue
 		}
 		n := map[string]envoy_types.ResourceWithTTL{}
+		changed = true
 		for k, v := range new.GetResourcesAndTTL(typ) {
 			n[k] = v
 		}
```
Comment (jakubdyszkiewicz) [inline pkg/kds/v2/reconcile/reconciler.go:127]: IMHO the content of for should look like this
```
	for _, typ := range util_kds_v2.GetSupportedTypes() {
		version := new.GetVersion(typ)
		if version == "" {
			if old != nil && r.equal(new.GetResources(typ), old.GetResources(typ)) {
				version = old.GetVersion(typ)
			} else {
				version = core.NewUUID()
			}
		}

		newResources[core_model.ResourceType(typ)] = envoy_cache.Resources{
			Version: version,
			Items:   new.GetResourcesAndTTL(typ), // do we really need to clone this?
		}

		if old.GetVersion(typ) != new.GetVersion(typ) {
			changed = true
		}
	}
```

**Comment by slonka** [comment]
Do not merge, this could potentially be reconciling every time.

**Comment by slonka** [comment]
> 2024-03-29T16:45:25.1274422Z 2024-03-29T16:45:25.126Z	INFO	kds-delta-global	received NACK will retry	{"nodeID": "kuma-1:", "type": "MeshCircuitBreaker", "err": "failed to create k8s resource: admission webhook \"owner-reference.kuma-admission.kuma.io\" denied the request: Mesh.kuma.io \"default\" not found", "delay": "5s"}

failed due to mesh not being created for 60 seconds :| I think this indicates a problem with checking the version changed

File: pkg/kds/v2/server/event_based_watchdog_test.go [hunk 2]
```diff
@@ -21,6 +21,9 @@ type staticReconciler struct {
 	changedResTypes chan map[core_model.ResourceType]struct{}
 }
 
+func (s staticReconciler) ForceVersion(node *envoy_core.Node, resourceType core_model.ResourceType) {
+}
```
Comment (slonka) [inline pkg/kds/v2/server/event_based_watchdog_test.go:25]: why is this empty?

File: pkg/kds/v2/reconcile/reconciler.go [hunk 3]
```diff
@@ -88,79 +96,67 @@ func (r *reconciler) Reconcile(ctx context.Context, node *envoy_core.Node, chang
...
+	if old != nil {
+		// this should already be computed by SetSnapshot, but we call it just to make sure we have versions.
+		if err := old.ConstructVersionMap(); err != nil {
+			return errors.Wrap(err, "could not construct version map"), false
+		}
+	}
+	if err := new.ConstructVersionMap(); err != nil {
+		return errors.Wrap(err, "could not construct version map"), false
+	}
+	r.forceNewVersion(new, id)
```
[the remark this answers is not in the document]
Comment (slonka) [reply pkg/kds/v2/reconcile/reconciler.go:109]: this is used to populate things for `changedTypes`

File: pkg/kds/v2/server/event_based_watchdog_test.go (hunk 2 above)
Comment (slonka) [reply pkg/kds/v2/server/event_based_watchdog_test.go:25]: this is for tests only now

**Review by slonka — COMMENTED** [review]
thanks for adding a test 👍

**Review by jakubdyszkiewicz — APPROVED** [review]

**Review by Automaat — APPROVED** [review]

**Pull request merged**